### DISCUSSION TITLE
What to do when file size not available to the VFS up front?

### DISCUSSION THREAD

**Discussion by tdgvrr** [dCache/nfs4j#106] [Q&A]
Apologies if this is a newbie question - still figuring out the nitty-gritty details of NFS as well as the nfs4j implementation.

It seems nfs4j normally requires the server to know the size of files being accessed before the file data is fetched...when a client reads a file, the server seems to get the size, then calls the VFS layer to fetch precisely this much data.  

Is there an alternative where the server can just read until it gets EOF from the VFS implementation? This is very handy when the underlying object isn't a filesystem object, but rather a pipe, a database query or some other "blob" where it's difficult or time-consuming to know the precise size involved ahead of time. What we're looking for (perhaps) is a way to set the size attribute to something like -1 (or not return it at all), then throw some sort of EOF exception to tell the system that a particular filehandle reached the end of its data.

**Comment by kofemann** [comment]
This is of course not a nice solution for NFS clients, as then try to optimize IO requests based on the file size....

You can always return Long.MAX_VALUE ans the file size, and return less bytes and set EOF flag on read:

```java
bool eof = false;
final READ4res res = result.opread;

int n;
try {
  n = vfs.read(...);
  res
} catch (EOFException e) {
  res.resok4.eof = true;
} 
```
You will need to implement a custom READ operation to use it:

```java
   class CustomNFSv4OperationFactory extends MDSOperationExecutor {

        @Override
        protected AbstractNFSv4Operation getOperation(nfs_argop4 op) {

            switch (op.argop) {
                case nfs_opnum4.OP_READ:
                    return new CustomOperationREAD(op);
                default:
                    return super.getOperation(op);
```

**Reply by tdgvrr** [reply]
I know it's been a while, but I wanted to circle back to this...

I did as you suggested earlier - for certain requests, we return MAX_VALUE as the file size, then in our custom READ operation, we set the res.resok4.eof = true flag exactly as you show. 

This doesn't seem to work - the client (which is just a standard NFS 4.1 client from a Ubuntu Linux system) just keeps coming back for more reads until it consumes MAX_VALUE bytes. Is there a "res.status" value we also need to set? Something special with the buffer?  

I'm sure we're doing something wrong - any idea what else we might need to look at?

**Reply by kofemann** [reply]
I will assume, that the client issues a `GETATTR` call to verify the `EOF` marker. Can you after reaching the end-of-file provide the real filesize on getattr?

**Reply by tdgvrr** [reply]
Once the file is opened, I'm not seeing any further GETATTR calls for the file. During OPEN, we pass Long.MAX_VALUE as the file size (as was suggested in the earlier post), then we don't see another GETATTR until the file is closed. 

We have a test that runs a database query that returns 4118 bytes. It's triggered by the client mounting our filesystem, then doing something like "cat test".  

The flow we see on the server looks like this: 

```
...
OP_OPEN
OP_GETFH (several attributes are fetched - size is returned as Long.MAX_VALUE)
OP_ACCESS
OP_GETATTR
```

Once the file is opened, we see the data being read:

```
OP_SEQUENCE
OP_PUTFH
OP_READ
```

We're given a large (512K) buffer, so the entire output fits in this first buffer. We return the full 4118 bytes, we set res.resok4.eof = true and res.status = nfsstat.NFS_OK. On the client, this data is received properly. But instead of detecting the EOF, the client comes back to us with another read:

```
OP_SEQUENCE
OP_PUTFH
OP_READ
```

We detect that this is a read after EOF, so this one returns an empty buffer, with res.resok4.eof = true and res.status = nfsstat.NFS_OK. The pattern of SEQUENCE/PUTFH/READ  just loops indefinitely at this point. There are no other operations - no GETATTR, etc - until we break out of the client program and then we see the file being closed. It's as though the client ignores the EOF marker we send and it's just looping until it gets Long.MAX_VALUE bytes returned.   

We're thinking to run a network trace on the client to verify that the EOF flag is being set properly, but otherwise, we're at a loss.

Any ideas you have are definitely appreciated!

**Reply by kofemann** [reply]
Hm.. Can you provide packet capture (tcpdum -i /tmp/nfs.pca port 2049)?

**Reply by tdgvrr** [reply]
Sure - here you go...it's the same scenario as above, but this time the data we're sending is 4518 bytes.

I see this in Wireshark - note the eof = 1 flag:
```
Frame 12: 1814 bytes on wire (14512 bits), 1814 bytes captured (14512 bits)
Ethernet II, Src: Dell_5f:74:cb (54:bf:64:5f:74:cb), Dst: EliteGro_6f:c0:0f (1c:69:7a:6f:c0:0f)
Internet Protocol Version 4, Src: 192.168.1.250, Dst: 192.168.1.201
Transmission Control Protocol, Src Port: 2049, Dst Port: 961, Seq: 3513, Ack: 873, Len: 1748
Remote Procedure Call, Type:Reply XID:0xc768dc9a
    Fragment header: Last fragment, 4624 bytes
    XID: 0xc768dc9a (3345538202)
    Message Type: Reply (1)
    [Program: NFS (100003)]
    [Program Version: 4]
    [Procedure: COMPOUND (1)]
    Reply State: accepted (0)
    [This is a reply to a request in frame 8]
    [Time from request: 0.244896000 seconds]
    Verifier
    Accept State: RPC executed successfully (0)
Network File System, Ops(3): SEQUENCE PUTFH READ
    [Program Version: 4]
    [V4 Procedure: COMPOUND (1)]
    Status: NFS4_OK (0)
    Tag: <EMPTY>
    Operations (count: 3)
        Opcode: SEQUENCE (53)
        Opcode: PUTFH (22)
            Status: NFS4_OK (0)
        Opcode: READ (25)
            Status: NFS4_OK (0)
            eof: 1
            Read length: 4518
            Data: <DATA>
                length: 4518
                contents: <DATA>
                fill bytes: opaque data
    [Main Opcode: READ (25)]
```
After this, we see the same sequence of OP_READ requests as I described earlier - EOF is still set, but there's no data length or buffer being returned:

```
Network File System, Ops(3): SEQUENCE PUTFH READ
    [Program Version: 4]
    [V4 Procedure: COMPOUND (1)]
    Status: NFS4_OK (0)
    Tag: <EMPTY>
    Operations (count: 3)
        Opcode: SEQUENCE (53)
        Opcode: PUTFH (22)
            Status: NFS4_OK (0)
        Opcode: READ (25)
            Status: NFS4_OK (0)
            eof: 1
            Read length: 0
            Data: <EMPTY>
    [Main Opcode: READ (25)]
```
If it's important, this is Ubuntu 22.04, but I believe we had the same issue under 18.04 and 20.04 as well. 

Thanks for your help!

[tcpdump.zip](https://github.com/dCache/nfs4j/files/11149572/tcpdump.zip)

**Reply by kofemann** [reply]
Ok, so here is what is happening. As your file is quite big (according to the size), the nfs client issues multiple READ requests in parallel. Some of them (one) get the data, and the others get zero bytes + EOF. As expected (see above) the client gets confused, and issues GETATTR (frames 144-154):

```
 $ tshark -r nfs.dmp -Y "nfs && frame.number in {144..154}"
  144   3.053683 192.168.1.250 → 192.168.1.201 NFS V4 Reply (Call In 140) READ
  145   3.072729 192.168.1.250 → 192.168.1.201 NFS V4 Reply (Call In 142) READ
  147   3.097270 192.168.1.201 → 192.168.1.250 NFS V4 Call GETATTR FH: 0x78128b77
  148   3.183426 192.168.1.250 → 192.168.1.201 NFS V4 Reply (Call In 147) GETATTR
  149   3.187413 192.168.1.201 → 192.168.1.250 NFS V4 Call READ StateID: 0xae0f Offset: 25165824 Len: 1048576
  150   3.232190 192.168.1.250 → 192.168.1.201 NFS V4 Reply (Call In 149) READ
  151   3.233908 192.168.1.201 → 192.168.1.250 NFS V4 Call READ StateID: 0xae0f Offset: 26214400 Len: 524288
  152   3.234630 192.168.1.201 → 192.168.1.250 NFS V4 Call READ StateID: 0xae0f Offset: 26738688 Len: 262144
  154   3.290765 192.168.1.250 → 192.168.1.201 NFS V4 Reply (Call In 151) READ

```

As the file use is still reported as MAX_VALUE

```
$ tshark -r nfs.dmp -Y "frame.number == 148" -Tfields -e nfs.fattr4.size
9223372036854775807
``

the client keeps reading ... 😎

**Reply by tdgvrr** [reply]
Thank you very much for this...I'm no slouch, but it would have taken me a long time to figure out that the NFS client was sending parallel requests for the same file, but that does make sense. 

I tried a quick test to verify this is indeed what's happening. Since I know my test data is 4-5K, instead of returning Long.MAX_VALUE for the file size, I just returned something bigger than my data can be. When I returned 32K as the length, the client seems to be behaving properly now. 

I'll enhance my code to track of the actual size of the dynamic output we're generating and then return this value on GETATTR once the dynamic output is finalized. From what I'm seeing, that should work fine. 

Thanks again for your help!
